### PR TITLE
Add ninja proof-of-concept

### PR DESCRIPTION
**Pull request by stefanv**
`python generate-ninja.py` -> create build.ninja
`ninja ../python-book/filename.html` -> build that filename ninja python-book -> build entire  python book
`ninja clean` -> clean up

`ninja -t clean` -> clean up files generated by ninja during build process

### CODE REVIEW

**Comment by stefanv** [comment]
@matthew-brett Want to play around with this a bit and see whether it's an improvement? If not, we should just stick to what we have.

**Comment by stefanv** [comment]
I can't seem to get `probability_theory_4_finite.Rmd` built on my system, but everything else works.

**Comment by matthew-brett** [comment]
What's the problem with `probability_theory_4_finite`?

**Comment by stefanv** [comment]
Huh, I think quarto may not be safe to run in parallel processes :thinking:

**Comment by stefanv** [comment]
```
WARNING: Error removing file /home/stefan/p/resampling/resampling-with/source/site_libs/quarto-html: No such file or directory (os error 2): remove '/home/stefan/p/resampling/resampling-with/source/site_libs/quarto-html'
ERROR: WalkError: No such file or directory (os error 2): readdir '/home/stefan/p/resampling/resampling-with/source/site_libs/bootstrap' for path "/home/stefan/p/resampling/resampling-with/source/site_libs/bootstrap"

WalkError: No such file or directory (os error 2): readdir '/home/stefan/p/resampling/resampling-with/source/site_libs/bootstrap' for path "/home/stefan/p/resampling/resampling-with/source/site_libs/bootstrap"
    at wrapErrorWithPath (file:///home/stefan/.local/quarto-1.4.76/bin/quarto.js:2201:12)
    at walkSync (file:///home/stefan/.local/quarto-1.4.76/bin/quarto.js:2218:15)
    at walkSync.next (<anonymous>)
    at copyMinimal (file:///home/stefan/.local/quarto-1.4.76/bin/quarto.js:36946:16)
    at renderProject (file:///home/stefan/.local/quarto-1.4.76/bin/quarto.js:66135:29)
    at eventLoopTick (ext:core/01_core.js:181:11)
    at async Command.fn (file:///home/stefan/.local/quarto-1.4.76/bin/quarto.js:70094:32)
    at async Command.execute (file:///home/stefan/.local/quarto-1.4.76/bin/quarto.js:8104:13)
    at async quarto (file:///home/stefan/.local/quarto-1.4.76/bin/quarto.js:102125:5)
    at async file:///home/stefan/.local/quarto-1.4.76/bin/quarto.js:102143:9
    ```

**Comment by stefanv** [comment]
And here's that chapter that won't compile:

```
[1/1] quarto render probability_theory_4_finite.Rmd --to html --profile python
FAILED: ../python-book/probability_theory_4_finite.html
quarto render probability_theory_4_finite.Rmd --to html --profile python


processing file: probability_theory_4_finite.Rmd

output file: probability_theory_4_finite.knit.md

pandoc
  to: html
  output-file: probability_theory_4_finite.html
  standalone: true
  toc: true
  number-sections: true
  title-prefix: Resampling statistics
  section-divs: true
  html-math-method: mathjax
  wrap: none
  default-image-extension: png
  css:
    - style.css
    - font-awesome.min.css
  number-offset:
    - 11

metadata
  crossref:
    chapters: true
    sec-prefix: Section
    fig-prefix: Figure
    tbl-prefix: Table
  document-css: false
  link-citations: true
  date-format: long
  lang: en
  bibliography:
    - simon_refs.bib
  knitr_settings:
    out_comment: ''
  noteout:
    pre-filter:
      - todo
      - comment
    filter-divspans:
      - r
    nb-flatten-divspans:
      - +
      - python
    nb-format: ipynb
    strip-header-nos: false
    binder-url: 'https://mybinder.org/v2/gh/resampling-stats/resampling-with/gh-pages?filepath=python-book/'
    nb-dir: notebooks
  theme: cosmo
  jupyter:
    jupytext:
      metadata_filter:
        notebook:
          additional: all
          excluded:
            - language_info
      text_representation:
        extension: .Rmd
        format_name: rmarkdown
        format_version: '1.0'
        jupytext_version: 0.8.6
    kernelspec:
      display_name: Python 3
      language: python
      name: python3
  resampling_with:
    ed2_fname: 12-Chap-8
  hide-description: true
  title: "[[11]{.chapter-number}\_ [Probability Theory, Part 4: Estimating Probabilities from Finite Universes]{.chapter-title}]{#sec-finite-universes .quarto-section-identifier}"

Error running Lua:
...cal/quarto-1.4.76/share/pandoc/datadir/lpegshortcode.lua:111: attempt to concatenate a nil value
stack traceback:
	[C]: in function 'lpeg.match'
	...stefan/.local/quarto-1.4.76/share/filters/qmd-reader.lua:116: in function 'parse_shortcodes'
	...stefan/.local/quarto-1.4.76/share/filters/qmd-reader.lua:133: in function 'Reader'
ninja: build stopped: subcommand failed.
```

**Comment by matthew-brett** [comment]
Can you replicate with quarto render on that file?

**Comment by stefanv** [comment]
This should be reasonably close to ready, if you want to take a look @matthew-brett.

**Comment by stefanv** [comment]
Unfortunately, I often have to run `ninja python-book` twice, because of a race condition in Quarto: https://github.com/quarto-dev/quarto-cli/issues/6116

**Comment by matthew-brett** [comment]
What do you think about a `source/Makefile` with:

```
build.ninja: generate-ninja.py
	python3 generate-ninja.py
	echo "Now run ninja to make targets"
```

?

File: README.md [hunk 1]
```diff
@@ -159,21 +143,15 @@ Editing:
...
-      cannot open the connection
-    Calls: .main ... eval -> <Anonymous> -> yaml.load -> paste -> readLines
+    ninja ../python-book/reliability_average.html
     ```
-
-    then make sure you have run the `make set-version-python` or `make
-    set-version-r` command, to build the needed `_quarto.yml` and
-    `_variables.yml` files.
+    You can see the generated contents by opening e.g.
+    `../python-book/reliability_average.Rmd`
```
Comment (matthew-brett) [inline README.md:152]: Does the build actually generate `.Rmd` files?

File: source/.ninja-usage [hunk 2]
```diff
@@ -0,0 +1,42 @@
+\e[1m\e[33mUsage: \e[97mninja [\e[1;36mOPTIONS\e[97m] [\e[36mTARGETS...\e[97m]\e[0m
```
Comment (matthew-brett) [inline source/.ninja-usage:1]: I see these stray characters displayed as-is in my terminal - what are they?   Color markers?

File: README.md (hunk 1 above)
Comment (stefanv) [reply README.md:152]: No, you're right, only HTML.

File: source/.ninja-usage (hunk 2 above)
Comment (stefanv) [reply source/.ninja-usage:1]: Yes, odd that your terminal does not support ansi! It's not like that's a very new standard 🤔 We can take it out. It's pretty though :)

**Comment by stefanv** [comment]
> Is there a way to add a conditional rule for the PDF builds, that encodes the dependency of the generated `.png` files on the matching `.svg` files?

Sure, that's straightforward.

File: source/.ninja-usage (hunk 2 above)
Comment (matthew-brett) [reply source/.ninja-usage:1]: It looks like the problem is with the literal `\e` to mean the Escape char in the `.ninja-usage` text.  If I replace `\e` with the actual code for escape character  - ^[ - I get the correct output.

Comment (matthew-brett) [reply source/.ninja-usage:1]: This gives the right output on my system:

```python
w.rule('print-help', r"cat .ninja-usage | sed 's/\\e/^[/g'")
```

Comment (stefanv) [reply source/.ninja-usage:1]: I changed `\e` to `\033`; maybe that works.

**Comment by stefanv** [comment]
> Is there a way to add a conditional rule for the PDF builds, that encodes the dependency of the generated `.png` files on the matching `.svg` files?

I've updated the rule to add that dependency, and to use the non-parallel build by default (since parallel builds are not supported).

**Comment by stefanv** [comment]
Hrm, looks like PDF build requires PDFs, not PNGs.

**Comment by stefanv** [comment]
OK, that should take care of it.

**Comment by stefanv** [comment]
OK, builds are running, it's just the status update that's failing.

**Comment by matthew-brett** [comment]
Is this one ready to merge?

**Comment by stefanv** [comment]
I will think a bit about how to fix the status update (it's never worked for PRs, as far as I know). The rest, I think, is ready if it works for you.

**Pull request merged**